text/0000-newtype.md

```diff
@@ -0,0 +1,360 @@
+and `Neg` is implemented for `U`, then `Neg` shall automatically be implemented
+for `T` via casting to `U`.
+
+For all types `W`, `Shl<W>` can be implemented for `T`. If for any integer type
+`I` `Shl<I>` is not is not explicitly implemented for `T` and `Shl<I>` is
+implemented for `U`, then `Shl<I>` shall automatically be implemented for `T`
+via casting to `U`.
+
+`T` implements `Drop` if and only if `U` implements `Drop`.
+
+`T` implements `Send` if and only if `U` implements `Send`.
+
+`T` implements `Sized` if and only if `U` implements `Sized`.
+
+`T` implements `Copy` if and only if `U` implements `Copy`.
+
+`T` implements `Sync` if and only if `U` implements `Sync`.
+
+For all types `W`, `Fn<W>` can be implemented for `T`.
+
+For all types `W`, `Index<W>` can be implemented for `T`.
+
+For all types `W`, `IndexMut<W>` can be implemented for `T`.
+
+`Deref` can be implemented for `T`.
+
+`DerefMut` can be implemented for `T`.
+
+The same rules as for `PartialOrd` apply to `PartialEq`.
+
+The same rules as for `PartialOrd` apply to `Add`.
+
+The same rules as for `PartialOrd` apply to `Sub`.
+
+The same rules as for `PartialOrd` apply to `Mul`.
+
+The same rules as for `PartialOrd` apply to `Div`.
+
+The same rules as for `PartialOrd` apply to `Rem`.
+
+The same rules as for `Neg` apply to `Not`.
+
+The same rules as for `PartialOrd` apply to `BitAnd`.
+
+The same rules as for `PartialOrd` apply to `BitOr`.
+
+The same rules as for `PartialOrd` apply to `BitXor`.
+
+The same rules as for `Shl` apply to `Shr`.
+
+The same rules as for `Fn` apply to `FnMut`.
+
+The same rules as for `Fn` apply to `FnOnce`.
+
+#### Default impls
+
+Let `Trait` be a trait with a default impl:
+
+```rust
+impl Trait for .. { }
+```
+
+If there is neither an explicit `impl Trait for T` nor an explicit negative
+`impl !Trait for T`, then `T` implements `Trait` if and only if `U` implements
+`Trait`.
+
+## Example
+
+Assume
+
+```rust
+newtype c_long = i64;
+```
+
+Then `c_long` has no inherent methods and implements the following operators:
+`==, <, >, <=, >=, <<, >>, |, &, ^, +, -, *, /, %`.
+
+`c_long` does not implement `Drop` but implements `Neg`, `Send`, `Sized`,
+`Copy`, and `Sync`.
+
+`c_long` does not coerce to `i64` but can be cast to any integer or floating
+point type. Conversely, any such type can be cast to `c_long`.
+
+`&c_long` does not coerce to `&i64` but can be cast to `&i64`.
+
+## Rationale
+
+The motivation of this RFC is to make `c_char`, `c_ulong`, `pid_t`, etc.
+distinct types that need to be explicitly cast at the "rustic" interface
+boundary or between each other. At first one might consider the following
+design:
+
+>`T` behaves as if it had been declared by `struct T(U);` except that one can
+>cast between `T` and `U`.
+
+There are some problems with this:
+
+- `U` has to be sized for the definition of `T` to make sense. 
+- `T` doesn't necessarily have the same memory representation as `U`.
```

Comment:
There's nothing sensible for the compiler to do to break it. This would be something I would be totally happy to guarantee, just like we guarantee the behaviour of `Option` sometimes.